File: backend_samples/library/javascript/AzureAIVisionFaceDeviceAttestation/tests/cbor.test.ts
import assert from 'node:assert/strict';
import { test } from 'node:test';
import { decodeCbor } from '../src/ios/cbor';
import { parseAppAttestAssertion, parseAppAttestToken, parseAssertionAuthData } from '../src/ios/parsers';
import { parseAttestAuthData } from '../src/ios/attestation_phases';
import { IosPhaseFail } from '../src/ios/types';
import { encode } from 'cbor';

test('App Attest header preserves unsigned counters and fixed field boundaries', () => {
  for (const counter of [0, 0x01020304, 0x7fffffff, 0x80000000, 0xffffffff]) {
    const authData = Buffer.alloc(38, 0xff);
    authData.fill(0x42, 0, 32);
    authData.writeUInt32BE(counter, 33);
    assert.deepEqual(parseAssertionAuthData(authData), {
      rpIdHash: Buffer.alloc(32, 0x42), flags: 255, signCount: counter,
    });
  }
  assert.throws(() => parseAssertionAuthData(Buffer.alloc(36)), {
    message: 'authenticatorData is 36 bytes, < 37',
  });
});

test('App Attest credential layout preserves lengths and truncation errors', () => {
  for (const length of [0, 1, 255, 256, 32768, 65535]) {
    const authData = Buffer.alloc(55 + length + 1, 0xee);
    authData.fill(0x42, 0, 32);
    authData.fill(0xff, 32, 37);
    authData.fill(0x61, 37, 53);
    authData.writeUInt16BE(length, 53);
    authData.fill(0xab, 55, 55 + length);
    assert.deepEqual(parseAttestAuthData(authData), {
      rpIdHash: Buffer.alloc(32, 0x42), flags: 255, signCount: 4294967295,
      aaguid: Buffer.alloc(16, 0x61), credIdLen: length, credentialId: Buffer.alloc(length, 0xab),
    });
  }
  for (const length of [0, 36, 37, 54, 55]) {
    const authData = Buffer.alloc(length);
    if (length === 55) authData[54] = 1;
    const failure = parseAttestAuthData(authData);
    assert.ok(failure instanceof IosPhaseFail);
    assert.equal(failure.message, length < 37 ? `authData is ${length} bytes, < 37`
      : length < 55 ? 'authData missing attested credential data' : 'authData truncated within credentialId');
  }
});

test('CBOR preserves supported types and offsets', () => {
  for (const [hex, expected] of [
    ['00', 0], ['17', 23], ['182a', 42], ['190100', 256], ['1a00010000', 65536],
    ['1b0000000100000000', 4294967296], ['29', -10], ['43010203', Buffer.from([1, 2, 3])],
    ['63616263', 'abc'], ['820102', [1, 2]], ['a20102616103', new Map<string | number, number>([[1, 2], ['a', 3]])],
  ] as const) {
    const bytes = Buffer.from(hex, 'hex');
    assert.deepEqual(decodeCbor(Buffer.concat([Buffer.from([0]), bytes, Buffer.from([0])]), 1), [expected, bytes.length + 1]);
  }
});

test('CBOR rejects malformed and unsupported encodings', () => {
  for (const hex of ['', '18', '430102', '636162', '8201', 'a16161', '9f01ff', 'bf616101ff',
    '5f4101ff', '7f6161ff', 'c001', 'f5', 'f6', 'f93c00', 'a18001', '1bffffffffffffffff', 'a26161f5616101']) {
    assert.throws(() => decodeCbor(Buffer.from(hex, 'hex')), hex);
  }
});

test('CBOR enforces nesting limits', () => {
  decodeCbor(Buffer.concat([Buffer.alloc(16, 0x81), Buffer.from([0])]));
  assert.throws(() => decodeCbor(Buffer.concat([Buffer.alloc(17, 0x81), Buffer.from([0])])));
});

test('CBOR library output works with App Attest callers', () => {
  const authData = Buffer.alloc(37, 1);
  const certificate = Buffer.from([1, 2, 3]);
  const token = Buffer.from(encode({
    fmt: 'apple-appattest', authData,
    attStmt: { x5c: [certificate, certificate], receipt: Buffer.from([4]) },
  })).toString('base64');
  assert.deepEqual(parseAppAttestToken(token).credCertDer, certificate);
  assert.deepEqual(parseAppAttestToken(token).authData, authData);
  const assertion = Buffer.from(encode({ signature: certificate, authenticatorData: authData })).toString('base64');
  assert.deepEqual(parseAppAttestAssertion(assertion).signature, certificate);
});

test('App Attest schema rejects malformed maps and required fields', () => {
  const certificate = Buffer.from([1, 2, 3]);
  const valid = { fmt: 'apple-appattest', authData: Buffer.alloc(37), attStmt: { x5c: [certificate, certificate] } };
  for (const input of [
    [], 'text', 1, {},
    { ...valid, fmt: 1 }, { ...valid, authData: 'bytes' },
    { ...valid, attStmt: [] }, { ...valid, attStmt: {} },
    { ...valid, attStmt: { x5c: [certificate] } },
    { ...valid, attStmt: { x5c: ['bytes', certificate] } },
    { ...valid, attStmt: { x5c: [certificate, 'bytes'] } },
  ]) {
    assert.throws(() => parseAppAttestToken(Buffer.from(encode(input)).toString('base64')));
  }
});

test('App Attest schema preserves buffers, optional receipts, and extra-field tolerance', () => {
  const certificate = Buffer.from([0, 255, 128]);
  const authData = Buffer.from([255, 0, 128]);
  for (const receipt of [undefined, 'ignored', Buffer.from([255, 0])]) {
    const attStmt = { x5c: [certificate, certificate, 'ignored'], ...(receipt === undefined ? {} : { receipt }) };
    const input = { fmt: 'policy-checked-later', authData, attStmt, extra: 1 };
    const result = parseAppAttestToken(Buffer.from(encode(input)).toString('base64'));
    assert.deepEqual(result, {
      fmt: input.fmt, authData, credCertDer: certificate, intermediateDer: certificate,
      receipt: Buffer.isBuffer(receipt) ? receipt : null,
      receiptLength: Buffer.isBuffer(receipt) ? receipt.length : 0,
    });
  }
});

test('App assertion schema enforces byte strings and minimum authenticator length', () => {
  const signature = Buffer.from([255, 0, 128]);
  const authenticatorData = Buffer.alloc(37, 255);
  for (const input of [
    [], 'text', {}, { signature }, { authenticatorData },
    { signature: 'bytes', authenticatorData }, { signature, authenticatorData: 'bytes' },
    { signature, authenticatorData: Buffer.alloc(36) },
  ]) {
    assert.throws(() => parseAppAttestAssertion(Buffer.from(encode(input)).toString('base64')));
  }
  for (const length of [37, 38]) {
    const input = { signature, authenticatorData: Buffer.alloc(length, 255) };
    assert.deepEqual(parseAppAttestAssertion(Buffer.from(encode({ ...input, extra: 1 })).toString('base64')), input);
  }
});

File: backend_samples/library/java/AzureAIVisionFaceDeviceAttestation/src/test/java/com/azure/ai/vision/face/deviceattestation/ios/CborTest.java
package com.azure.ai.vision.face.deviceattestation.ios;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.Arrays;
import java.util.Base64;
import java.util.HexFormat;
import java.io.ByteArrayOutputStream;
import java.lang.reflect.InvocationTargetException;
import com.azure.ai.vision.face.deviceattestation.Json;
import com.fasterxml.jackson.dataformat.cbor.CBORFactory;
import java.util.List;
import java.util.Map;

import org.junit.jupiter.api.Test;

class CborTest {

    @Test
    void attestedCredentialLayoutPreservesFieldsAndBoundaryErrors() throws Exception {
        var parser = AppAttestVerification.class.getDeclaredMethod("parseAttestAuthData", byte[].class);
        parser.setAccessible(true);
        for (int length : new int[] { 0, 1, 255, 256, 32768, 65535 }) {
            byte[] authData = new byte[55 + length + 1];
            Arrays.fill(authData, 0, 32, (byte) 0x42);
            Arrays.fill(authData, 32, 37, (byte) 0xff);
            Arrays.fill(authData, 37, 53, (byte) 0x61);
            authData[53] = (byte) (length >>> 8);
            authData[54] = (byte) length;
            Arrays.fill(authData, 55, 55 + length, (byte) 0xab);
            authData[55 + length] = (byte) 0xee;
            Object parsed = parser.invoke(null, (Object) authData);
            assertArrayEquals(Arrays.copyOfRange(authData, 0, 32), (byte[]) component(parsed, "rpIdHash"));
            assertEquals(255, component(parsed, "flags"));
            assertEquals(4294967295L, component(parsed, "signCount"));
            assertArrayEquals(Arrays.copyOfRange(authData, 37, 53), (byte[]) component(parsed, "aaguid"));
            assertEquals(length, component(parsed, "credIdLen"));
            assertArrayEquals(Arrays.copyOfRange(authData, 55, 55 + length), (byte[]) component(parsed, "credentialId"));
        }
        for (int length : new int[] { 0, 36, 37, 54, 55 }) {
            byte[] truncated = new byte[length];
            if (length == 55) truncated[54] = 1;
            var error = assertThrows(InvocationTargetException.class, () -> parser.invoke(null, (Object) truncated));
            String expected = length < 37 ? "authData is " + length + " bytes, < 37"
                    : length < 55 ? "authData missing attested credential data" : "authData truncated within credentialId";
            assertEquals(expected, error.getCause().getMessage());
        }
    }

    @Test
    void envelopeRecordPreservesTokenAndAssertionPairing() throws Exception {
        byte[] authData = new byte[55];
        byte[] assertionData = new byte[37];
        assertionData[36] = 1;
        byte[] signature = { 1, 2, 3 };
        var tokenOutput = new ByteArrayOutputStream();
        try (var writer = new CBORFactory().createGenerator(tokenOutput)) {
            writer.writeStartObject(null, 3);
            writer.writeStringField("fmt", "apple-appattest");
            writer.writeFieldName("authData");
            writer.writeBinary(authData);
            writer.writeFieldName("attStmt");
            writer.writeStartObject(null, 1);
            writer.writeFieldName("x5c");
            writer.writeStartArray(null, 2);
            writer.writeBinary(new byte[] { 4 });
            writer.writeBinary(new byte[] { 5 });
            writer.writeEndArray();
            writer.writeEndObject();
            writer.writeEndObject();
        }
        var assertionOutput = new ByteArrayOutputStream();
        try (var writer = new CBORFactory().createGenerator(assertionOutput)) {
            writer.writeStartObject(null, 2);
            writer.writeFieldName("signature");
            writer.writeBinary(signature);
            writer.writeFieldName("authenticatorData");
            writer.writeBinary(assertionData);
            writer.writeEndObject();
        }
        var parser = AppAttestVerification.class.getDeclaredMethod("parseEnvelope", String.class);
        parser.setAccessible(true);
        Object envelope = parser.invoke(null, Json.stringify(Map.of(
                "attestation", Base64.getEncoder().encodeToString(tokenOutput.toByteArray()),
                "assertion", Base64.getEncoder().encodeToString(assertionOutput.toByteArray()))));
        var token = assertInstanceOf(AppAttestObject.class, component(envelope, "token"));
        var assertion = assertInstanceOf(AppAttestAssertionObject.class, component(envelope, "assertion"));
        assertEquals("apple-appattest", token.fmt);
        assertArrayEquals(authData, token.authData);
        assertArrayEquals(new byte[] { 4 }, token.credCertDer);
        assertArrayEquals(new byte[] { 5 }, token.intermediateDer);
        assertArrayEquals(signature, assertion.signature);
        assertArrayEquals(assertionData, assertion.authenticatorData);
        var error = assertThrows(InvocationTargetException.class, () -> parser.invoke(null, "{}"));
        assertEquals("Invalid attestation: missing attestation field", error.getCause().getMessage());
    }

    private static Object component(Object record, String name) throws Exception {
        var accessor = record.getClass().getDeclaredMethod(name);
        accessor.setAccessible(true);
        return accessor.invoke(record);
    }

    @Test
    void assertionCountersRemainUnsignedAndBigEndian() {
        String[] encodings = { "00000000", "01020304", "7fffffff", "80000000", "ffffffff" };
        long[] expected = { 0L, 0x01020304L, 2147483647L, 2147483648L, 4294967295L };
        for (int index = 0; index < encodings.length; index++) {
            byte[] authData = new byte[38];
            Arrays.fill(authData, 0, 32, (byte) 0x42);
            authData[32] = (byte) 0x80;
            System.arraycopy(HexFormat.of().parseHex(encodings[index]), 0, authData, 33, 4);
            authData[37] = (byte) 0xff;
            var parsed = AppAttestParsers.parseAssertionAuthData(authData);
            assertEquals(expected[index], parsed.signCount());
            assertEquals(128, parsed.flags());
            assertArrayEquals(Arrays.copyOfRange(authData, 0, 32), parsed.rpIdHash());
        }
        assertThrows(IllegalArgumentException.class,
                () -> AppAttestParsers.parseAssertionAuthData(new byte[36]));
    }

    @Test
    void libraryEncodingWorksWithAssertionParser() throws Exception {
        byte[] signature = { 1, 2, 3 };
        byte[] authData = new byte[37];
        var output = new ByteArrayOutputStream();
        try (var writer = new CBORFactory().createGenerator(output)) {
            writer.writeStartObject(null, 2);
            writer.writeFieldName("signature");
            writer.writeBinary(signature);
            writer.writeFieldName("authenticatorData");
            writer.writeBinary(authData);
            writer.writeEndObject();
        }
        var assertion = AppAttestParsers.parseAppAttestAssertion(Base64.getEncoder().encodeToString(output.toByteArray()));
        assertArrayEquals(signature, assertion.signature);
        assertArrayEquals(authData, assertion.authenticatorData);
    }

    @Test
    void rejectsMalformedOrUnsupportedValues() {
        for (String hex : List.of("", "18", "430102", "636162", "8201", "a16161", "9f01ff", "bf616101ff",
            "5f4101ff", "7f6161ff", "c001", "f5", "f6", "f93c00", "a18001", "1bffffffffffffffff",
            "a1416101", "a1c0616101", "a17f6161ff01")) {
            assertThrows(RuntimeException.class, () -> Cbor.decode(HexFormat.of().parseHex(hex)), hex);
        }
    }

    @Test
    void preservesOffsetsAndNumericKeys() {
        var decoded = Cbor.decode(HexFormat.of().parseHex("002900"), 1);
        assertEquals(-10L, decoded.value());
        assertEquals(2, decoded.nextPos());
        assertEquals(4294967296L, Cbor.decode(HexFormat.of().parseHex("1b0000000100000000")).value());
        assertEquals(Map.of(1L, 2L, "a", 3L), Cbor.decode(HexFormat.of().parseHex("a20102616103")).value());
        assertEquals(Map.of(-1L, 2L, 1L, 3L, "a", 4L),
            Cbor.decode(HexFormat.of().parseHex("a320020103616104")).value());
    }

    @Test
    void limitsNesting() {
        byte[] allowed = new byte[17];
        Arrays.fill(allowed, 0, 16, (byte) 0x81);
        Cbor.decode(allowed);
        byte[] excessive = new byte[18];
        Arrays.fill(excessive, 0, 17, (byte) 0x81);
        assertThrows(RuntimeException.class, () -> Cbor.decode(excessive));
    }

    @Test
    void decodesUnsignedInt() {
        assertEquals(10L, Cbor.decode(new byte[] { 0x0a }).value());
    }

    @Test
    void decodesTextString() {
        // 0x61 = text string length 1, 0x61 = 'a'
        assertEquals("a", Cbor.decode(new byte[] { 0x61, 0x61 }).value());
    }

    @Test
    void decodesByteString() {
        // 0x42 = byte string length 2
        assertArrayEquals(new byte[] { 0x01, 0x02 }, (byte[]) Cbor.decode(new byte[] { 0x42, 0x01, 0x02 }).value());
    }

    @Test
    void decodesArray() {
        // 0x83 = array of 3
        Object value = Cbor.decode(new byte[] { (byte) 0x83, 0x01, 0x02, 0x03 }).value();
        assertEquals(List.of(1L, 2L, 3L), value);
    }

    @Test
    @SuppressWarnings("unchecked")
    void decodesMap() {
        // 0xa1 = map of 1: "a" -> 1
        Object value = Cbor.decode(new byte[] { (byte) 0xa1, 0x61, 0x61, 0x01 }).value();
        assertInstanceOf(Map.class, value);
        Map<Object, Object> map = (Map<Object, Object>) value;
        assertEquals(1L, map.get("a"));
    }
}
